src/core/ext/filters/client_channel/subchannel.cc

```diff
@@ -717,19 +690,27 @@ channelz::SubchannelNode* Subchannel::channelz_node() {
 void Subchannel::WatchConnectivityState(
     const absl::optional<std::string>& health_check_service_name,
     RefCountedPtr<ConnectivityStateWatcherInterface> watcher) {
-  MutexLock lock(&mu_);
-  grpc_pollset_set* interested_parties = watcher->interested_parties();
-  if (interested_parties != nullptr) {
-    grpc_pollset_set_add_pollset_set(pollset_set_, interested_parties);
-  }
-  if (!health_check_service_name.has_value()) {
-    new AsyncWatcherNotifierLocked(watcher, state_, status_);
-    watcher_list_.AddWatcherLocked(std::move(watcher));
-  } else {
-    health_watcher_map_.AddWatcherLocked(
-        WeakRef(DEBUG_LOCATION, "health_watcher"), *health_check_service_name,
-        std::move(watcher));
+  {
+    MutexLock lock(&mu_);
+    grpc_pollset_set* interested_parties = watcher->interested_parties();
+    if (interested_parties != nullptr) {
+      grpc_pollset_set_add_pollset_set(pollset_set_, interested_parties);
+    }
+    if (!health_check_service_name.has_value()) {
+      work_serializer_.Schedule(
+          [watcher = watcher->Ref(), state = state_, status = status_]() {
+            watcher->OnConnectivityStateChange(state, status);
+          },
+          DEBUG_LOCATION);
+      watcher_list_.AddWatcherLocked(std::move(watcher));
+    } else {
+      health_watcher_map_.AddWatcherLocked(
+          WeakRef(DEBUG_LOCATION, "health_watcher"), *health_check_service_name,
+          std::move(watcher));
+    }
   }
+  // Drain any connectivity state notifications after releasing the mutex.
+  work_serializer_.DrainQueue();
```

Comment:
Also consider what happens if in `CancelConnectivityStateWatch` on line 716, we later add some WorkSerializer schedules in one of the `RemoveWatcherLocked` functions